sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,592 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Thomas Moreau <thomas.moreau.2010@gmail.com>
+# Author: Olivier Grisel <olivier.grisel@ensta.fr>
+
+
+from cpython cimport Py_INCREF, PyObject
+
+from libc.stdlib cimport malloc, free
+from libc.string cimport memcpy
+from libc.stdio cimport printf
+
+from sklearn.tree._utils cimport safe_realloc, sizet_ptr_to_ndarray
+from ..utils import check_array
+
+import numpy as np
+cimport numpy as np
+np.import_array()
+
+cdef extern from "math.h":
+    float fabsf(float x) nogil
+
+cdef extern from "numpy/arrayobject.h":
+    object PyArray_NewFromDescr(object subtype, np.dtype descr,
+                                int nd, np.npy_intp* dims,
+                                np.npy_intp* strides,
+                                void* data, int flags, object obj)
+
+
+# XXX using (size_t)(-1) is ugly, but SIZE_MAX is not available in C89
+# (i.e., older MSVC).
+cdef SIZE_t DEFAULT = <SIZE_t>(-1)
+
+
+# Repeat struct definition for numpy
+CELL_DTYPE = np.dtype({
+    'names': ['parent', 'children', 'cell_id', 'point_index', 'is_leaf',
+              'max_width', 'depth', 'cumulative_size', 'center', 'barycenter',
+              'min_bounds', 'max_bounds'],
+    'formats': [np.intp, (np.intp, 8), np.intp, np.intp, np.int32, np.float32,
+                np.intp, np.intp, (np.float32, 3), (np.float32, 3),
+                (np.float32, 3), (np.float32, 3)],
+    'offsets': [
+        <Py_ssize_t> &(<Cell*> NULL).parent,
+        <Py_ssize_t> &(<Cell*> NULL).children,
+        <Py_ssize_t> &(<Cell*> NULL).cell_id,
+        <Py_ssize_t> &(<Cell*> NULL).point_index,
+        <Py_ssize_t> &(<Cell*> NULL).is_leaf,
+        <Py_ssize_t> &(<Cell*> NULL).max_width,
+        <Py_ssize_t> &(<Cell*> NULL).depth,
+        <Py_ssize_t> &(<Cell*> NULL).cumulative_size,
+        <Py_ssize_t> &(<Cell*> NULL).center,
+        <Py_ssize_t> &(<Cell*> NULL).barycenter,
+        <Py_ssize_t> &(<Cell*> NULL).min_bounds,
+        <Py_ssize_t> &(<Cell*> NULL).max_bounds,
+    ]
+})
+
+assert CELL_DTYPE.itemsize == sizeof(Cell)
+
+
+cdef class QuadTree:
+    """Array-based representation of a QuadTree.
+    """
```

Comment:
Please add a note to state that the current code works both for QuadTree (4 children cells per level, suitable for indexing 2D data) and OcTree (8 children cells per level, suitable for indexing 3D data).

Please also add a note stating that this Cython code is currently internally used only by the Barnes Hut method of the `sklearn.manifold.TSNE` class and that it is planed to be slightly re factored and generalized in the future to be suitable as a backend method for `sklearn.neighbors.KNearestNeighbors` for 2D and 3D data.